src/openms/source/FILTERING/DATAREDUCTION/Deisotoper.cpp

```diff
@@ -38,9 +38,341 @@
 #include <OpenMS/FILTERING/DATAREDUCTION/Deisotoper.h>
 #include <OpenMS/KERNEL/MSSpectrum.h>
 #include <OpenMS/MATH/MISC/MathFunctions.h>
+#include <OpenMS/CHEMISTRY/ISOTOPEDISTRIBUTION/CoarseIsotopePatternGenerator.h>
+#include <OpenMS/FILTERING/TRANSFORMERS/ThresholdMower.h>
+#include <OpenMS/FILTERING/TRANSFORMERS/NLargest.h>
 
 namespace OpenMS
 {
+// static
+void Deisotoper::deisotopeWithAveragineModel(MSSpectrum& spec,
+  double fragment_tolerance,
+  bool fragment_unit_ppm,
+  bool rem_low_intensity,
+  int min_charge,
+  int max_charge,
+  bool keep_only_deisotoped,
+  unsigned int min_isopeaks,
+  unsigned int max_isopeaks,
+  bool make_single_charged,
+  bool annotate_charge,
+  bool annotate_iso_peak_count,
+  bool add_up_intensity,
+  bool used_for_open_search)
+{
+  OPENMS_PRECONDITION(spec.isSorted(), "Spectrum must be sorted.");
+
+  if (min_isopeaks < 2 || max_isopeaks < 2 || min_isopeaks > max_isopeaks)
+  {
+    throw Exception::IllegalArgument(__FILE__,
+      __LINE__,
+      OPENMS_PRETTY_FUNCTION,
+      "Minimum/maximum number of isotopic peaks must be at least 2 (and min_isopeaks <= max_isopeaks).");
+  }
+
+  if (spec.empty()) { return; }
+
+  // remove 0 intensity peaks
+  ThresholdMower threshold_mower_filter;
+  threshold_mower_filter.filterPeakSpectrum(spec);
+
+  // discard low-intensity peaks
+  if (rem_low_intensity)
+  { 
+    UInt max_num_peaks = used_for_open_search ? 1000 : 5000;
+
+    // only keep max_num_peaks highest peaks
+    NLargest nlargest_filter = NLargest(max_num_peaks);
```

Comment:
are you sure this is actually a good idea to filter the spectrum BEFORE finding isotope peaks? You might trim a isotope clusters.
Do you have some data on the effect of the filtering?

Reply:
I think it is described in the publication like that.

Reply:
Yes, this part, and also the numbers, are taken directly from the publication. I think most of my spectrums do not even have 5000 peaks but I will investigate with those that do in 2 weeks